cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java

```diff
@@ -156,53 +156,48 @@ public ProgramController run(final Program program, ProgramOptions options) {
                                                                         program.getJarLocation(), locationFactory,
                                                                         streamAdmin, txSystemClient);
-    final ProgramController controller = new MapReduceProgramController(mapReduceRuntimeService, context);
-    controller.addListener(new AbstractListener() {
+    mapReduceRuntimeService.addListener(new Service.Listener() {
       @Override
-      public void init(ProgramController.State state, @Nullable Throwable cause) {
+      public void starting() {
         store.setStart(program.getId(), runId.getId(), TimeUnit.SECONDS.convert(System.currentTimeMillis(),
                                                                                 TimeUnit.MILLISECONDS));
-        if (state == ProgramController.State.COMPLETED) {
-          completed();
-        }
-        if (state == ProgramController.State.ERROR) {
-          error(controller.getFailureCause());
-        }
       }
 
       @Override
-      public void completed() {
-        store.setStop(program.getId(), runId.getId(),
-                      TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS),
-                      ProgramController.State.COMPLETED.getRunStatus());
+      public void running() {
+        //no-op
       }
 
       @Override
-      public void killed() {
-        store.setStop(program.getId(), runId.getId(),
-                      TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS),
-                      ProgramController.State.KILLED.getRunStatus());
+      public void stopping(Service.State from) {
+        //no-op
       }
 
       @Override
-      public void suspended() {
-        store.setSuspend(program.getId(), runId.getId());
-      }
-
-      @Override
-      public void resuming() {
-        store.setResume(program.getId(), runId.getId());
+      public void terminated(Service.State from) {
+        if (from != Service.State.STOPPING) {
+          // Service completed by itself.
+          store.setStop(program.getId(), runId.getId(),
+                        TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS),
+                        ProgramController.State.COMPLETED.getRunStatus());
+        } else {
+          // Service was killed
+          store.setStop(program.getId(), runId.getId(),
+                        TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS),
+                        ProgramController.State.KILLED.getRunStatus());
+        }
       }
 
       @Override
-      public void error(Throwable cause) {
-        LOG.info("MapReduce Program stopped with error {}, {}", program.getId(), runId, cause);
+      public void failed(Service.State from, Throwable failure) {
         store.setStop(program.getId(), runId.getId(),
                       TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS),
```

Comment:
I find it simpler and easier to read if you use `TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())`